Update helm chart for cilium-operator to implement per-provider opera…

seanmwinn opened the pull request:
…tor deployment for generic, aws and azure

Fixes: #11800

Signed-off-by: Sean Winn <sean@isovalent.com>


```release-note
cilium-operator now packages a separate container for generic use cases, as well as additional container images for Microsoft Azure and Amazon Web Services. The default configuration will deploy the cilium-operator-generic image. For AWS, configure `.Values.global.eni=true` to deploy the cilium-operator-aws image. For Azure configure `.Values.global.azure.enabled=true` to deploy the cilium-operator-azure image. Only one cloud provider is supported in a Cilium deployment.
```

seanmwinn committed:
Implement per-provider operator deployments

Update the helm charts for for generic, aws and azure operator images.

Fixes: #11800

Signed-off-by: Sean Winn <sean@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

aanm commented on install/kubernetes/cilium/charts/operator/values.yaml:
```diff
@@ -1,4 +1,11 @@
+enabled: true
+
+image:
+  repository: cilium-operator
```
`s/cilium-operator/operator`

seanmwinn replied:
According to @errordeveloper this is supposed to change to the cilium-operator-generic container, so I am receiving conflicting information. This also uses part of the deployment manifest which selects a container image based on ipam mode.

joestringer replied:
Right now it's `operator` / `operator-generic`:

https://hub.docker.com/repository/docker/cilium/operator
https://hub.docker.com/repository/docker/cilium/operator-generic

aanm commented on install/kubernetes/cilium/charts/operator/values.yaml:
```diff
@@ -1,4 +1,11 @@
+
+image:
+  repository: cilium-operator
+  tag: latest
```
cc @joestringer do you thing the `examples/install/Makefile` will catch this in the release process?

joestringer replied:
Looks like we'll need to amend the Makefile to look at this file in addition to the global values.yaml. I can prep a patch for that.

joestringer replied:
Pushed.

aanm reviewed (changes requested):
@seanmwinn please also update the release process to have the `imagePullPolicy` and `tag` modified.

https://github.com/cilium/cilium/blob/6d52ae4cd0ffa5ddc29d29a8eb2b08d05e43913b/Documentation/contributing/release/stable.rst#L42-L51

joestringer committed:
install: Fix up version/pullPolicy for multiple values files

In particular, add the new operator values file to the list of files
that we fix these values up for in the make target.

Signed-off-by: Joe Stringer <joe@cilium.io>

errordeveloper commented on install/kubernetes/cilium/charts/operator/templates/deployment.yaml:
```diff
@@ -33,7 +33,13 @@ spec:
         - --debug=$(CILIUM_DEBUG)
         command:
-        - cilium-operator
+{{- if .Values.global.eni }}
```
The image should pick the correct binary as default command, but if we prefer this to be explicit I won't oppose :)

seanmwinn replied:
If neither of the provider flags is set, the default else condition will use cilium-operator-generic as the command

errordeveloper reviewed (approved):
LGTM overall! Thanks @seanmwinn!

aanm reviewed (changes requested):
@joestringer the release process guide needs to be updated, right?

sayboras commented on install/kubernetes/cilium/charts/operator/templates/deployment.yaml:
```diff
@@ -101,12 +107,30 @@ spec:
+{{- if .Values.global.pullPolicy }}
         imagePullPolicy: {{ .Values.global.pullPolicy }}
+{{- else }}
+        imagePullPolicy: {{ .Values.image.pullPolicy }}
```
should it be _.Values.image.imagePullPolicy_ instead ?

sayboras commented on install/kubernetes/cilium/charts/operator/values.yaml:
```diff
@@ -1,4 +1,8 @@
+image:
+  repository: operator
+  tag: latest
+  imagePullPolicy: Always
```
related to above comment, this field can be rename to just _pullPolicy_ (as .Values.image.imagePullPolicy looks a little bit weird I think).

sayboras commented on test/helpers/kubectl.go:
```diff
@@ -91,7 +91,7 @@ var (
 		"preflight.image":                     "cilium-dev", // Set again in init to match agent.image!
 		"global.tag":                          "latest",
-		"operator.image":                      "operator",
+		"operator.image.repository":           "operator-generic",
```
any case we have two _generic_ here ? https://github.com/cilium/cilium/pull/11837/files#diff-58754dc409d3caca0cbf3bb702b551d3R124

aanm commented on Documentation/contributing/testing/e2e.rst:
```diff
@@ -513,7 +513,7 @@ cluster.
 ::
 
-  CNI_INTEGRATION=eks K8S_VERSION=1.14 ginkgo -v --focus="K8sDemo*" -noColor -- -cilium.provision=false -cilium.kubeconfig=`echo ~/.kube/config` -cilium.image="docker.io/cilium/cilium:latest" -cilium.operator-image="docker.io/cilium/operator:latest" -cilium.passCLIEnvironment=true
+  CNI_INTEGRATION=eks K8S_VERSION=1.14 ginkgo -v --focus="K8sDemo*" -noColor -- -cilium.provision=false -cilium.kubeconfig=`echo ~/.kube/config` -cilium.image="docker.io/cilium/cilium:latest" -cilium.operator-image="docker.io/cilium/operator-generic:latest" -cilium.passCLIEnvironment=true
```
`s/-generic/-aws`?

seanmwinn replied:
fixed

aanm commented on jenkinsfiles/ginkgo-eks.Jenkinsfile:
```diff
@@ -125,7 +125,7 @@ pipeline {
             steps {
                 dir("${TESTDIR}"){
-                    sh 'CILIUM_IMAGE=$(./print-node-ip.sh)/cilium/cilium:latest CILIUM_OPERATOR_IMAGE=$(./print-node-ip.sh)/cilium/operator:latest ginkgo --focus="$(echo ${ghprbCommentBody} | sed -r "s/([^ ]* |^[^ ]*$)//" | sed "s/^$/K8s*/")" -v --failFast=${FAILFAST} -- -cilium.provision=false -cilium.timeout=${GINKGO_TIMEOUT} -cilium.kubeconfig=${TESTDIR}/eks/eks-kubeconfig -cilium.passCLIEnvir ...
+                    sh 'CILIUM_IMAGE=$(./print-node-ip.sh)/cilium/cilium:latest CILIUM_OPERATOR_IMAGE=$(./print-node-ip.sh)/cilium/operator-generic:latest ginkgo --focus="$(echo ${ghprbCommentBody} | sed -r "s/([^ ]* |^[^ ]*$)//" | sed "s/^$/K8s*/")" -v --failFast=${FAILFAST} -- -cilium.provision=false -cilium.timeout=${GINKGO_TIMEOUT} -cilium.kubeconfig=${TESTDIR}/eks/eks-kubeconfig -cilium.pass ...
```
`s/-generic/-aws`?

seanmwinn replied:
fixed

aanm commented on Documentation/contributing/testing/e2e.rst:
```diff
@@ -513,7 +517,11 @@ cluster.
   CNI_INTEGRATION=eks K8S_VERSION=1.14 ginkgo -v --focus="K8sDemo*" -noColor -- -cilium.provision=false -cilium.kubeconfig=`echo ~/.kube/config` -cilium.image="docker.io/cilium/cilium:latest" -cilium.operator-image="docker.io/cilium/operator:latest" -cilium.passCLIEnvironment=true
+=======
+  CNI_INTEGRATION=eks K8S_VERSION=1.14 ginkgo -v --focus="K8sDemo*" -noColor -- -cilium.provision=false -cilium.kubeconfig=`echo ~/.kube/config` -cilium.image="docker.io/cilium/cilium:latest" -cilium.operator-image="docker.io/cilium/operator-aws:latest" -cilium.passCLIEnvironment=true
+>>>>>>> e27b6ef06... fix up tests for aws operator image
```
Bad rebase? :-)

aanm commented on jenkinsfiles/ginkgo-eks.Jenkinsfile:
```diff
@@ -125,7 +125,11 @@ pipeline {
                     sh 'CILIUM_IMAGE=$(./print-node-ip.sh)/cilium/cilium:latest CILIUM_OPERATOR_IMAGE=$(./print-node-ip.sh)/cilium/operator:latest ginkgo --focus="$(echo ${ghprbCommentBody} | sed -r "s/([^ ]* |^[^ ]*$)//" | sed "s/^$/K8s*/")" -v --failFast=${FAILFAST} -- -cilium.provision=false -cilium.timeout=${GINKGO_TIMEOUT} -cilium.kubeconfig=${TESTDIR}/eks/eks-kubeconfig -cilium.passCLIEnvir ...
+=======
+                    sh 'CILIUM_IMAGE=$(./print-node-ip.sh)/cilium/cilium:latest CILIUM_OPERATOR_IMAGE=$(./print-node-ip.sh)/cilium/operator-aws:latest ginkgo --focus="$(echo ${ghprbCommentBody} | sed -r "s/([^ ]* |^[^ ]*$)//" | sed "s/^$/K8s*/")" -v --failFast=${FAILFAST} -- -cilium.provision=false -cilium.timeout=${GINKGO_TIMEOUT} -cilium.kubeconfig=${TESTDIR}/eks/eks-kubeconfig -cilium.passCLIE ...
+>>>>>>> e27b6ef06... fix up tests for aws operator image
```
Bad rebase? :-)

seanmwinn committed:
fix rendering for image names with a / to not append suffixes

Signed-off-by: Sean Winn <sean@isovalent.com>

joestringer commented on install/kubernetes/Makefile:
```diff
@@ -8,15 +8,15 @@ MANAGED_ETCD_VERSION := "v2.0.7"
-CILIUM_CHARTS := "$(ROOT_DIR)/$(RELATIVE_DIR)/cilium/"
-CILIUM_VALUES := "$(CILIUM_CHARTS)/values.yaml"
+CILIUM_CHARTS := "$(ROOT_DIR)/$(RELATIVE_DIR)/cilium"
+CILIUM_VALUES := "$(CILIUM_CHARTS)/values.yaml" "$(CILIUM_CHARTS)/charts/operator/values.yaml"
```
When rebasing, this change will still be necessary.

joestringer commented on install/kubernetes/cilium/charts/operator/values.yaml:
```diff
@@ -1,4 +1,8 @@
-image: operator
+image:
+  repository: operator
+  tag: latest
+  pullPolicy: Always
+
```
All of the other changes were already applied in other PRs, except for this change.

joestringer replied:
(Well, and obviously when this changes the actual daemonset needs to take this into account)

joestringer replied:
We will need to look at this closely from an upgrade instructions angle because changing the "image" value from a string to a map may cause helm to reject the upgrade.

joestringer reviewed (changes requested):
Quick status update to clarify what was merged / what is deferred to resolve in the future.
